restart/monitor/manager.go

```diff
@@ -0,0 +1,223 @@
+}
+
+func init() {
+	plugin.Register(&plugin.Registration{
+		Type: plugin.InternalPlugin,
+		Requires: []plugin.Type{
+			plugin.ServicePlugin,
+		},
+		ID: "restart",
+		Config: &Config{
+			Interval: duration{
+				Duration: 10 * time.Second,
+			},
+		},
+		InitFn: func(ic *plugin.InitContext) (interface{}, error) {
+			opts, err := getServicesOpts(ic)
+			if err != nil {
+				return nil, err
+			}
+			client, err := containerd.New("", containerd.WithServices(opts...))
+			if err != nil {
+				return nil, err
+			}
+			m := &monitor{
+				client: client,
+			}
+			go m.run(ic.Config.(*Config).Interval.Duration)
+			return m, nil
+		},
+	})
+}
+
+// getServicesOpts get service options from plugin context.
+func getServicesOpts(ic *plugin.InitContext) ([]containerd.ServicesOpt, error) {
+	plugins, err := ic.GetByType(plugin.ServicePlugin)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get service plugin")
+	}
+	opts := []containerd.ServicesOpt{
+		containerd.WithEventService(ic.Events),
+	}
+	for s, fn := range map[string]func(interface{}) containerd.ServicesOpt{
+		services.ContentService: func(s interface{}) containerd.ServicesOpt {
+			return containerd.WithContentStore(s.(content.Store))
+		},
+		services.ImagesService: func(s interface{}) containerd.ServicesOpt {
+			return containerd.WithImageService(s.(images.ImagesClient))
+		},
+		services.SnapshotsService: func(s interface{}) containerd.ServicesOpt {
+			return containerd.WithSnapshotters(s.(map[string]snapshots.Snapshotter))
+		},
+		services.ContainersService: func(s interface{}) containerd.ServicesOpt {
+			return containerd.WithContainerService(s.(containers.ContainersClient))
+		},
+		services.TasksService: func(s interface{}) containerd.ServicesOpt {
+			return containerd.WithTaskService(s.(tasks.TasksClient))
+		},
+		services.DiffService: func(s interface{}) containerd.ServicesOpt {
+			return containerd.WithDiffService(s.(diff.DiffClient))
+		},
+		services.NamespacesService: func(s interface{}) containerd.ServicesOpt {
+			return containerd.WithNamespaceService(s.(namespacesapi.NamespacesClient))
+		},
+		services.LeasesService: func(s interface{}) containerd.ServicesOpt {
+			return containerd.WithLeasesService(s.(leases.LeasesClient))
+		},
+	} {
+		p := plugins[s]
+		if p == nil {
+			return nil, errors.Errorf("service %q not found", s)
+		}
+		i, err := p.Instance()
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to get instance of service %q", s)
+		}
+		if i == nil {
+			return nil, errors.Errorf("instance of service %q not found", s)
+		}
+		opts = append(opts, fn(i))
+	}
+	return opts, nil
+}
+
+type change interface {
+	apply(context.Context, *containerd.Client) error
+}
+
+type monitor struct {
+	client *containerd.Client
+}
+
+func (m *monitor) run(interval time.Duration) {
+	if interval == 0 {
+		interval = 10 * time.Second
+	}
+	for range time.Tick(interval) {
+		logrus.Debug("running restart monitor reconcile loop")
```

Comment:
Can we get rid of this debug line? A debug line after desired state isn't matched is probably more useful.